### PR TITLE
persist: restore fetch_and_update_state after upper mismatch in CaA

### PR DESCRIPTION
**Pull request by pH14**
The intended behavior after hitting an upper mismatch during `compare_and_append` is to fetch the latest state and compare the upper to see if the batch's lower still matches. Fetching the latest state was erroneously removed in https://github.com/MaterializeInc/materialize/pull/18488

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by danhhz — COMMENTED** [review]
any chance I could talk you into writing a regression test? feel free to bail if it ends up being terrible

**Comment by pH14** [comment]
> any chance I could talk you into writing a regression test? feel free to bail if it ends up being terrible

Wasn't bad at all! Thanks for the nudge, I initially thought it was going to be hard to write to the point of not being worthwhile.

File: src/persist-client/src/internal/machine.rs [hunk 1]
```diff
@@ -2002,4 +2004,39 @@ pub mod tests {
...
+            (("3".to_owned(), ()), 3, 1),
+            (("4".to_owned(), ()), 4, 1),
+            (("5".to_owned(), ()), 5, 1),
+        ];
+
+        write1.expect_compare_and_append(&data[..1], 0, 2).await;
+        // quick check: each handle should have its own copy of state
+        assert!(write1.machine.seqno() > write2.machine.seqno());
+        // this handle's upper now lags behind. if compare_and_append fails to update state
+        // after an upper mismatch we would expect this call to fail
```
Comment (danhhz) [inline src/persist-client/src/internal/machine.rs:2039]: ultra nit: s/after an upper mismatch we would expect this call to fail/after an upper mismatch then this call would (incorrectly) fail/

**Review by danhhz — APPROVED** [review]

**Review by bkirwi — APPROVED** [review]
🌟

**Pull request merged**